### PR TITLE
Update table of required image formats (#427)

### PR DESCRIPTION
**Pull request by Jeremy-Kemp**
This commit updates the table of required image formats. The table is built depending on the profile of the device, the requested image type, and the avaiablitiy of relevent extensions.

### CODE REVIEW

**Comment by Jeremy-Kemp** [comment]
This is a pre-requisite change for #624

File: test_common/harness/imageHelpers.cpp [hunk 1]
```diff
@@ -3671,45 +3671,201 @@ bool find_format( cl_image_format *formatList, unsigned int numFormats, cl_image
...
+		{ CL_RGBA, CL_FLOAT },
+	};
+
+	/*
+		Required full profile formats.
+		This array does not contain any full profile
+		formats that have restrictions on when they
+		are required.
+	*/
+	cl_image_format fullProfReadOrWriteFormats[] =
```
Comment (kpet) [inline test_common/harness/imageHelpers.cpp]: This list is not valid for an OpenCL 1.2 implementation (see https://github.com/KhronosGroup/OpenCL-Docs/issues/229).

File: test_common/harness/imageHelpers.cpp [hunk 2]
```diff
@@ -3671,45 +3671,201 @@ bool find_format( cl_image_format *formatList, unsigned int numFormats, cl_image
...
-        { CL_BGRA, CL_UNORM_INT8} };
+bool check_minimum_supported(cl_image_format *formatList,
+                             unsigned int numFormats,
+                             cl_mem_flags flags,
+                             cl_mem_object_type image_type,
+                             cl_device_id device)
+{
+	bool passed = true;
+	bool bDepth = false;
+	bool bsRGBA = false;
```
Comment (kpet) [inline test_common/harness/imageHelpers.cpp]: These could maybe have a more explicit name. I'm thinking something like `depth_required` (or better).

File: test_common/harness/imageHelpers.cpp [hunk 3]
```diff
@@ -3671,45 +3671,201 @@ bool find_format( cl_image_format *formatList, unsigned int numFormats, cl_image
...
+			uiTotalFormatSize += sizeof(fullProfReadOrWriteDepthFormats);
+			bDepth = true;
+		}
+
+		// sRGB writes not required for 1DImage Buffers
+		if (image_type != CL_MEM_OBJECT_IMAGE1D_BUFFER)
+		{
+			if (flags == CL_MEM_WRITE_ONLY)
+			{
+				// sRGB writes are only required with the cl_khr_srgb_image_writes extension
```
Comment (kpet) [inline test_common/harness/imageHelpers.cpp]: I think that's incorrect. Supporting the extension doesn't mean that the formats become required.

File: test_common/harness/imageHelpers.cpp [hunk 4]
```diff
@@ -3671,45 +3671,201 @@ bool find_format( cl_image_format *formatList, unsigned int numFormats, cl_image
...
+	}
+
+	cl_image_format *readFormatsToSupport = (cl_image_format*)malloc(uiTotalFormatSize);
+	cl_image_format *writeFormatsToSupport = (cl_image_format*)malloc(uiTotalFormatSize);
+
+	if (readFormatsToSupport == NULL || writeFormatsToSupport == NULL)
+	{
+		passed = false;
+		goto exit;
+	}
```
Comment (kpet) [inline test_common/harness/imageHelpers.cpp]: I would suggest to use `std::vector` to simplify that code.

File: test_common/harness/imageHelpers.cpp [hunk 5]
```diff
@@ -3671,45 +3671,201 @@ bool find_format( cl_image_format *formatList, unsigned int numFormats, cl_image
...
+        testCount = uiTotalFormatSize / sizeof( readFormatsToSupport[ 0 ] );
     }
     else
     {
         formatsToTest = writeFormatsToSupport;
-        testCount = sizeof( writeFormatsToSupport ) / sizeof( writeFormatsToSupport[ 0 ] );
+        testCount = uiTotalFormatSize / sizeof( writeFormatsToSupport[ 0 ] );
     }
 
     for( unsigned int i = 0; i < testCount; i++ )
```
Comment (kpet) [inline test_common/harness/imageHelpers.cpp]: The message below needs to be updated with the actual version supported by the device.

File: test_common/harness/imageHelpers.cpp [hunk 6]
```diff
@@ -3671,45 +3671,201 @@ bool find_format( cl_image_format *formatList, unsigned int numFormats, cl_image
...
+	}
+	else
+	{
+		size_t uiReadOffset = 0;
+		size_t uiWriteOffset = 0;
+
+		memcpy(readFormatsToSupport + uiReadOffset,
+		       fullProfReadOrWriteFormats,
+		       sizeof(fullProfReadOrWriteFormats));
+		uiReadOffset += sizeof(fullProfReadOrWriteFormats) / sizeof(fullProfReadOrWriteFormats[0]);
```
Comment (kpet) [inline test_common/harness/imageHelpers.cpp]: You should use `ARRAY_SIZE`.

File: test_common/harness/imageHelpers.cpp [hunk 7]
```diff
@@ -3671,45 +3671,201 @@ bool find_format( cl_image_format *formatList, unsigned int numFormats, cl_image
...
+                             cl_mem_flags flags,
+                             cl_mem_object_type image_type,
+                             cl_device_id device)
+{
+	bool passed = true;
+	bool bDepth = false;
+	bool bsRGBA = false;
+
+	// Required embedded formats.
+	cl_image_format embeddedProfReadOrWriteFormats[] =
```
Comment (kpet) [inline test_common/harness/imageHelpers.cpp]: All these tables should be `static`.

File: test_common/harness/imageHelpers.cpp (hunk 2 above)
Comment (Jeremy-Kemp) [reply test_common/harness/imageHelpers.cpp]: Are the new names OK?

File: test_common/harness/imageHelpers.cpp [hunk 8]
```diff
@@ -3671,56 +3674,173 @@ bool find_format( cl_image_format *formatList, unsigned int numFormats, cl_image
...
+		{ CL_sRGBA, CL_UNORM_INT8 },
+	};
+
+	static std::vector<cl_image_format> fullProfReadOrWriteDepthFormats
+	{
+		{ CL_DEPTH, CL_UNORM_INT16 },
+		{ CL_DEPTH, CL_FLOAT },
+	};
+
+	static std::vector<cl_image_format> fullProfSRGBFormats
```
Comment (Jeremy-Kemp) [inline test_common/harness/imageHelpers.cpp]: This needs to be update based on the outcome of [KhronosGroup/OpenCL-Docs/issues/229](https://github.com/KhronosGroup/OpenCL-Docs/issues/229).

Comment (Jeremy-Kemp) [reply test_common/harness/imageHelpers.cpp]: I've made sRGB match the layout of sRGBA for now.

File: test_common/harness/imageHelpers.cpp (hunk 3 above)
Comment (Jeremy-Kemp) [reply test_common/harness/imageHelpers.cpp]: Agree, I mis-read the spec. Updated to only require `CL_MEM_READ_ONLY`.

File: test_common/harness/imageHelpers.cpp (hunk 1 above)
Comment (Jeremy-Kemp) [reply test_common/harness/imageHelpers.cpp]: Should be correct now based on the version of the device.

File: test_common/harness/imageHelpers.cpp [hunk 9]
```diff
@@ -3671,56 +3677,181 @@ bool find_format( cl_image_format *formatList, unsigned int numFormats, cl_image
...
+		(specifically 2.x).
+		There are cases whereby the format isn't required.
+	*/
+	static std::vector<cl_image_format> fullProf2XSRGBFormats
+	{
+		{CL_sRGB, CL_UNORM_INT8},
+	};
+
+	// Full profile
+	if (!gIsEmbedded)
```
Comment (kpet) [inline test_common/harness/imageHelpers.cpp]: You could reverse the condition for clarity.

File: test_common/harness/imageHelpers.cpp [hunk 10]
```diff
@@ -3671,56 +3677,181 @@ bool find_format( cl_image_format *formatList, unsigned int numFormats, cl_image
...
+	// Embedded profile
+	else
+	{
+		copy(embeddedProfReadOrWriteFormats.begin(),
+		     embeddedProfReadOrWriteFormats.end(),
+		     back_inserter(readFormatsToSupport));
+
+		copy(embeddedProfReadOrWriteFormats.begin(),
+		     embeddedProfReadOrWriteFormats.end(),
+		     back_inserter(writeFormatsToSupport));
```
Comment (kpet) [inline test_common/harness/imageHelpers.cpp]: You could de-duplicate the copy code and assign a pointer depending on the profile.

Comment (Jeremy-Kemp) [reply test_common/harness/imageHelpers.cpp]: I've taken this further. Given that we're currently only looking at read _or_ write, and that the read or write formats are nearly identical, we can just add everything and special case what we need.

**Pull request merged**